### ISSUE TITLE
[Accepted] SDL 0306 - Use Taskmaster To Handle Queuing Operations in Managers

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#1021]
Hello SDL community,

The review of "SDL 0306 - Use Taskmaster To Handle Queuing Operations in Managers"  begins now and runs through June 2, 2020.  The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0306-Use-Taskmaster-To-Handle-Queuing-Operations-In-Managers.md

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/1021

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:

* Is the problem being addressed significant enough to warrant a change to SDL?
* Does this proposal fit well with the feel and direction of SDL?
* If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
* How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by kshala-ford** [comment]
I'm OK with this change. To my understanding every manager can create a queue while the master runs all of them on a single thread. Also the thread pool executor can run a pool of threads per instance allowing concurrent tasks per manager and each manager runs their own queue concurrently from other managers.

Are we expecting possible performance degradation as only one task can run at a time? When more managers use the taskmaster I could see one adding a longer running task and other manager's RPCs are not being send.

And sorry for asking a stupid question: Why not share a `ThreadPoolExecutor` object in the internal interface?

**Comment by joeygrover** [comment]
@kshala-ford  Not a stupid question at all, I was surprised to find that there really weren't any supporting types of threading tools natively available that met our needs and we did start off using executors/pools for one manager, but as the proposal mentions, extending this to adding independent executors/pools per manager is generally a poor choice. 

The `Taskmaster` essentially creates independent queues before the executor's single queue; this ensures that each the new manager queues has their tasks executed in a fairly distributed way. Without this mechanism, as managers add their runnable to lets say a single `ThreadPoolExecutor`, all those tasks will get added to a single queue. This means no other managers tasks will operate until the batch of runnable tasks completes. If `n'` long running tasks are added by a single manager where `n` is the number of threads allowable for the `ThreadPoolExecutor` to create, then performance severely degrades. Therefore, by adding queues before this single queue we enable other tasks to be executed in a more "round-robin" sort of way. It also allows queues to be created that are independently asynchronous or serial. Using one `ThreadPoolExecutor ` would only allow all task to be on a single, serial queue (poor performance). Additionally, each manager having their own independent queues allows them to traverse the queues to find outdated tasks and cancel them before their execution where possible, or simply getting status of where in their operations they are.

While it is possible to configure the `Taskmaster` with the single thread for executing tasks, in practice it is more likely to be a couple at max (~3-4); the underlying classes will handle the spinning up and down of those threads to the max level supplied by the SDL library. This will take some investigation to determine best performance and will likely change based on the number of managers that move over to a queue based system vs monolithic callback hells.

**Comment by bmcisaa1-ford** [comment]
Nice solution 👍

**Comment by kshala-ford** [comment]
Thanks @joeygrover this clarifies certain aspects of the proposal. Externals (non-PM) who donate code will need to learn how to use this library but can learn this from how it'll be used in the choice set manager.

From the technical perspective I want to provide you feedback to make the logger non-static (see [here](https://github.com/livio/Taskmaster/blob/master/Taskmaster/src/main/java/com/livio/taskmaster/TaskmasterLogger.java#L6)) or to rethink how the taskmaster can be build. I know you want to simplify calls to log by using e.g. `TaskmasterLogger.v(...)`  but when I tested the taskmaster I realized that only one taskmaster object can be build using the `Builder`. Also expecting proper unit tests to be created. There are a few tests in the repository however only one is executed which is probably due to the static logger.

I assume the project is owned and will be owned by the Livio Inc. and you propose to use this library in sdl_java_suite which is owned by the SDLC. I couldn't find any licensing conditions in the other repository. Can the PM elaborate on how you plan to license the code to the SDLC? Are the PM planning to fork the repository from Livio to smartdevicelink just like you do for other 3rd party repositories?

**Comment by joeygrover** [comment]
@kshala-ford I will continue to add more examples to the README of the library to help others as well. Though it should be very intuitive after one manager implements it.

I agree, the way the logger was being used didn't work for multiple Taskmasters. I opened a PR to change that. The unit tests right now are a little cluttered in the Test class that uses multiple static variables for testing which is why the tests haven't been ran in parallel, only one at a time but the logger issue would have also prevented that. I plan to refactor the test to be better scoped and easier to run.

I have added a license to the develop branch of BSD3 and would agree that this has to be on master before inclusion into the SDL project. With this being owned by the project maintainer and covered under a BSD3 license in its current and future state it will be free for the SDLC to use in perpetuity. We have forked other 3rd party repos in the past, though I'm not sure it is necessary here since the project is owned by the PM and can be forked at any time. The SDL Java Suite project will also be using the bintray published released so even if it is forked, the actual release will come from the upstream repo. However, if it is desired to create a fork into the SDL project I am fine with that.
